bots/elasticsearch/tasks.py

```diff
@@ -132,3 +138,79 @@ def serialize(self, source):
             'name': safe_substr(source.long_title),
             'short_name': safe_substr(source.name)
         }
+
+
+def check_counts_in_range(self, min_date, max_date):
+    partial_database_count = CreativeWork.objects.exclude(title='').exclude(is_deleted=True).filter(
+        date_created__range=[min_date, max_date]
+    ).count()
+    partial_es_count = self.es_client.count(
+        index=(self.es_index or settings.ELASTICSEARCH_INDEX),
+        doc_type='creativeworks',
+        body={
+            'query': {
+                'range': {
+                    'date_modified': {
+                        'gte': '{}'.format(min_date.isoformat()),
+                        'lte': '{}'.format(max_date.isoformat())
+                    }
+                }
+            }
+        }
+    )['count']
+    return (partial_database_count == partial_es_count, partial_database_count, partial_es_count)
+
+
+def get_date_range_parts(min_date, max_date):
+    middle_date = min_date.average(max_date)
+    return {
+        'first_half': {'min_date': min_date, 'max_date': middle_date},
+        'second_half': {'min_date': middle_date, 'max_date': max_date}
+    }
+
+
+@celery.task
+def pseudo_bisection_method(self, min_date, max_date):
+    '''
+    Checks if the database count matches the ES count
+    If counts differ, split the date range in half and check counts in smaller ranges
+    Pseudo binary because it can't throw away half the results based on the middle value
+    '''
+    from bots.elasticsearch.bot import ElasticSearchBot
+
+    MAX_DB_COUNT = 500
+    MIN_MISSING_RATIO = 0.7
+
+    counts_match, db_count, es_count = check_counts_in_range(self, min_date, max_date)
+
+    if counts_match:
+        return
+    if db_count <= MAX_DB_COUNT or 1 - abs(es_count / db_count) >= MIN_MISSING_RATIO:
+        logger.info('Counts for %s to %s do not match. %s creativeworks in ES, %s creativeworks in database.', min_date.format('%B %-d, %Y %I:%M:%S %p'), max_date.format('%B %-d, %Y %I:%M:%S %p'), es_count, db_count)
+        logger.info('Reindexing records created from %s to %s.', min_date.format('%B %-d, %Y %I:%M:%S %p'), max_date.format('%B %-d, %Y %I:%M:%S %p'))
+
+        bot = apps.get_app_config('elasticsearch').get_bot(
+            ShareUser.objects.get(username=settings.APPLICATION_USERNAME),
+            es_filter={'date_created__range': [min_date.isoformat(), max_date.isoformat()]},
+        )
+        bot.run()
```

Comment:
Should the task also remove records from the index which don't match a work in the database? I'm not sure if that's ever been a problem, but it seems possible.